File: markdown_writer.py
import os
import re
import logging

logger = logging.getLogger(__name__)


def _sanitize_filename(title):
    """Convert a recipe title to a safe filename."""
    name = title.lower().strip()
    name = re.sub(r"[^\w\s-]", "", name)
    name = re.sub(r"[\s_]+", "-", name)
    name = re.sub(r"-+", "-", name)
    return name[:80].strip("-")


def recipe_to_markdown(recipe):
    """Convert a recipe dict to a Markdown string."""
    lines = []

    # Title
    lines.append(f"# {recipe['title']}\n")

    # Description
    if recipe.get("description"):
        lines.append(f"{recipe['description']}\n")

    # Image
    if recipe.get("image"):
        lines.append(f"![{recipe['title']}]({recipe['image']})\n")

    # Time and servings table
    time_rows = []
    for label, key in [
        ("Prep Time", "prep_time"),
        ("Cook Time", "cook_time"),
        ("Total Time", "total_time"),
        ("Servings", "servings"),
    ]:
        value = recipe.get(key)
        if value:
            time_rows.append(f"| {label} | {value} |")

    if time_rows:
        lines.append("| | |")
        lines.append("|---|---|")
        lines.extend(time_rows)
        lines.append("")

    # Ingredients
    if recipe.get("ingredients"):
        lines.append("## Ingredients\n")
        for ing in recipe["ingredients"]:
            lines.append(f"- {ing}")
        lines.append("")

    # Instructions
    if recipe.get("instructions"):
        lines.append("## Instructions\n")
        step_num = 1
        for step in recipe["instructions"]:
            if step.startswith("**") and step.endswith("**"):
                # Section header
                lines.append(f"\n### {step.strip('*')}\n")
                step_num = 1
            else:
                lines.append(f"{step_num}. {step}")
                step_num += 1
        lines.append("")

    # Nutrition
    if recipe.get("nutrition"):
        lines.append("## Nutrition\n")
        lines.append("| Nutrient | Amount |")
        lines.append("|---|---|")
        for nutrient, amount in recipe["nutrition"].items():
            lines.append(f"| {nutrient} | {amount} |")
        lines.append("")

    # Footer metadata
    lines.append("---")
    if recipe.get("url"):
        lines.append(f"*Source: {recipe['url']}*  ")
    if recipe.get("categories"):
        lines.append(f"*Categories: {', '.join(recipe['categories'])}*  ")
    if recipe.get("cuisine"):
        lines.append(f"*Cuisine: {', '.join(recipe['cuisine'])}*  ")
    if recipe.get("keywords"):
        lines.append(f"*Keywords: {', '.join(recipe['keywords'])}*")
    lines.append("")

    return "\n".join(lines)


def save_recipe(recipe, output_dir):
    """Save a recipe as a Markdown file.

    Returns the file path on success, None on failure.
    """
    if not recipe or not recipe.get("title"):
        logger.warning("Cannot save recipe without a title")
        return None

    os.makedirs(output_dir, exist_ok=True)

    filename = _sanitize_filename(recipe["title"]) + ".md"
    filepath = os.path.join(output_dir, filename)

    md_content = recipe_to_markdown(recipe)

    with open(filepath, "w", encoding="utf-8") as f:
        f.write(md_content)

    logger.info(f"Saved: {filepath}")
    return filepath

File: android_src/app/src/main/kotlin/com/webscrape/recipemd/data/MarkdownWriter.kt
package com.webscrape.recipemd.data

import android.content.Context
import android.util.Log
import java.io.File

object MarkdownWriter {

    private const val TAG = "MarkdownWriter"

    fun toMarkdown(recipe: Recipe, localImagePath: String? = null): String = buildString {
        appendLine("# ${recipe.title}")
        appendLine()

        if (recipe.description.isNotEmpty()) {
            appendLine(recipe.description)
            appendLine()
        }

        val imageRef = localImagePath ?: recipe.imageUrl
        if (imageRef != null) {
            appendLine("![${recipe.title}]($imageRef)")
            appendLine()
        }

        // Time and servings table
        val timeRows = mutableListOf<Pair<String, String>>()
        recipe.prepTime?.let { timeRows.add("Prep Time" to it) }
        recipe.cookTime?.let { timeRows.add("Cook Time" to it) }
        recipe.totalTime?.let { timeRows.add("Total Time" to it) }
        recipe.servings?.let { timeRows.add("Servings" to it) }
        if (timeRows.isNotEmpty()) {
            appendLine("| | |")
            appendLine("|---|---|")
            for ((label, value) in timeRows) {
                appendLine("| $label | $value |")
            }
            appendLine()
        }

        // Ingredients
        if (recipe.ingredients.isNotEmpty()) {
            appendLine("## Ingredients")
            appendLine()
            for (ing in recipe.ingredients) {
                appendLine("- $ing")
            }
            appendLine()
        }

        // Instructions
        if (recipe.instructions.isNotEmpty()) {
            appendLine("## Instructions")
            appendLine()
            var stepNum = 1
            for (step in recipe.instructions) {
                if (step.startsWith("**") && step.endsWith("**")) {
                    appendLine()
                    appendLine("### ${step.trim('*')}")
                    appendLine()
                    stepNum = 1
                } else {
                    appendLine("$stepNum. $step")
                    stepNum++
                }
            }
            appendLine()
        }

        // Nutrition
        if (recipe.nutrition.isNotEmpty()) {
            appendLine("## Nutrition")
            appendLine()
            appendLine("| Nutrient | Amount |")
            appendLine("|---|---|")
            for ((nutrient, amount) in recipe.nutrition) {
                appendLine("| $nutrient | $amount |")
            }
            appendLine()
        }

        // Footer
        appendLine("---")
        if (recipe.url.isNotEmpty()) appendLine("*Source: ${recipe.url}*  ")
        if (recipe.categories.isNotEmpty()) appendLine("*Categories: ${recipe.categories.joinToString(", ")}*  ")
        if (recipe.cuisine.isNotEmpty()) appendLine("*Cuisine: ${recipe.cuisine.joinToString(", ")}*  ")
        if (recipe.keywords.isNotEmpty()) appendLine("*Keywords: ${recipe.keywords.joinToString(", ")}*")
    }

    fun saveRecipe(context: Context, recipe: Recipe, localImagePath: String? = null): File? {
        val dir = File(context.getExternalFilesDir(null), "recipes")
        dir.mkdirs()

        val filename = sanitizeFilename(recipe.title) + ".md"
        val file = File(dir, filename)

        return try {
            file.writeText(toMarkdown(recipe, localImagePath))
            Log.i(TAG, "Saved: ${file.absolutePath}")
            file
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save ${recipe.title}: ${e.message}")
            null
        }
    }

    fun saveImage(context: Context, imageBytes: ByteArray, title: String, imageUrl: String): File? {
        val dir = File(context.getExternalFilesDir(null), "recipes/images")
        dir.mkdirs()

        val extension = imageUrl.substringAfterLast('.', "jpg")
            .substringBefore('?')
            .lowercase()
            .let { if (it in listOf("jpg", "jpeg", "png", "webp", "gif")) it else "jpg" }
        val filename = sanitizeFilename(title) + ".$extension"
        val file = File(dir, filename)

        return try {
            file.writeBytes(imageBytes)
            Log.i(TAG, "Saved image: ${file.absolutePath}")
            file
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save image for $title: ${e.message}")
            null
        }
    }

    fun saveHtml(context: Context, html: String, title: String): File? {
        val dir = File(context.getExternalFilesDir(null), "recipes/html")
        dir.mkdirs()

        val filename = sanitizeFilename(title) + ".html"
        val file = File(dir, filename)

        return try {
            file.writeText(html)
            Log.i(TAG, "Saved HTML: ${file.absolutePath}")
            file
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save HTML for $title: ${e.message}")
            null
        }
    }

    private fun sanitizeFilename(title: String): String {
        return title.lowercase()
            .replace(Regex("[^\\w\\s-]"), "")
            .replace(Regex("[\\s_]+"), "-")
            .replace(Regex("-+"), "-")
            .take(80)
            .trim('-')
    }
}
